Add calibration-to-noise pipeline

95-martin-orion opened the pull request:
This PR reinstates the cirq_google utility for converting `Calibration` objects to noise properties. To create a simulator with behavior that mimics some `Calibration` "cal", the expected usage is:
```python
noise_props = cg.noise_properties_from_calibration(cal)
noise_model = cg.NoiseModelFromGoogleNoiseProperties(noise_props)
simulator = cirq.Simulator(noise=noise_model)  # will mimic noise from `calibration`
```

This is part 4 of #4666; part 3 was #5082. Remaining steps include:
- Convenience methods for modifying noise properties
- Reinstatement of compare_generated_noise_to_metrics using new types

MichaelBroughton commented on cirq-google/cirq_google/experimental/noise_models/calibration_to_noise_properties.py:
```diff
@@ -0,0 +1,164 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
Nit: module docstring.

95-martin-orion replied:
Done.

MichaelBroughton commented on cirq-google/cirq_google/experimental/noise_models/calibration_to_noise_properties.py:
```diff
@@ -0,0 +1,164 @@
+import numpy as np
+
+import cirq, cirq_google
+from cirq.devices import noise_utils
+from cirq_google.devices import google_noise_properties
```
Do we need these imports if we are doing broader imports just above ?

95-martin-orion replied:
Resolved...ish. See comment on circular imports :/

MichaelBroughton commented on cirq-google/cirq_google/experimental/noise_models/calibration_to_noise_properties.py:
```diff
@@ -0,0 +1,164 @@
+from typing import Dict, Tuple, Type
+import numpy as np
+
+import cirq, cirq_google
```
Nit: circular ?

95-martin-orion replied:
Turns out there's already a circular import in `google_noise_properties`:
https://github.com/quantumlib/Cirq/blob/e31b48d04c7a3dea6bc2e5e320b0c606bbf13700/cirq-google/cirq_google/devices/google_noise_properties.py#L22

More specifically, it uses `cirq_google.ops` (which comes after `cg.devices` in the hierarchy) for `SycamoreGate`. So I have some untangling to do. Possible side effects:
- Might remove public `*_QUBIT_GATES` sets in favor of in-method sets
- Probably moving calibration-to-noise out of experimental
- _Might_ have to move `google_noise_properties` as well (unlikely)

MichaelBroughton commented on cirq-google/cirq_google/experimental/noise_models/calibration_to_noise_properties.py:
```diff
@@ -0,0 +1,164 @@
+
+    # TODO: acquire this based on the target device.
+    # Default map of gates to their durations.
+    DEFAULT_GATE_NS: Dict[Type['cirq.Gate'], float] = {
```
Nit: snake case ?

95-martin-orion replied:
Done.

MichaelBroughton commented on cirq-google/cirq_google/experimental/noise_models/calibration_to_noise_properties.py:
```diff
@@ -0,0 +1,164 @@
+    if rb_incoherent_errors:
+        microwave_time_ns = DEFAULT_GATE_NS[cirq.PhasedXZGate]
+        for qubit, q_t1_ns in t1_ns.items():
+            tphi_err = rb_incoherent_errors[qubit] - microwave_time_ns / (3 * q_t1_ns)
+            if tphi_err > 0:
+                q_tphi_ns = microwave_time_ns / (3 * tphi_err)
+            else:
+                q_tphi_ns = 1e10
+            tphi_ns[qubit] = q_tphi_ns
```
```python3
q_tphi_ns = 1e10
if tphi_err > 0:
  q_tphi_ns = microwave_time_ns / (3 * tphi_err)
tphi_ns[qubit] = q_tphi_ns
```

95-martin-orion replied:
Done.

MichaelBroughton reviewed (approved):
LGTM after nits

95-martin-orion commented:
Removed the approval mostly to prevent myself from absentmindedly merging this before the circular dependencies are resolved - I'll ping reviewers when this is ready for re-review.

95-martin-orion commented:
Okay, I've moved these file to engine and removed circular imports. Imports are working, verified in another PR (which will go out separately from this and #5188).

This is ready for re-review.

MichaelBroughton commented on cirq-google/cirq_google/engine/calibration_to_noise_properties.py:
```diff
@@ -0,0 +1,178 @@
+Given a Calibration "cal", a user can simulate noise approximating that
+calibration using the following pipeline:
+
+    noise_props = cg.noise_properties_from_calibration(cal)
+    noise_model = cg.NoiseModelFromGoogleNoiseProperties(noise_props)
+    simulator = cirq.Simulator(noise=noise_model)
+    simulator.simulate(circuit)
```
Nit: Don't forget >>> to show command lines and put in output lines.

95-martin-orion replied:
Done, but we don't have a set standard for this - these rules are applied inconsistently throughout Cirq, and our docs don't mention them.

MichaelBroughton commented on cirq-google/cirq_google/engine/calibration_to_noise_properties.py:
```diff
@@ -0,0 +1,178 @@
+    To manually override noise properties, call `override` on the output:
+
+        # Set all gate durations to 37ns.
+        noise_properties_from_calibration(cal).override(gate_times_ns=37)
```
Nit: >>>

95-martin-orion replied:
Done, but see above.

MichaelBroughton commented on cirq-google/cirq_google/engine/calibration_to_noise_properties.py:
```diff
@@ -0,0 +1,178 @@
+    """Translates between a Calibration object and a NoiseProperties object.
+
+    The NoiseProperties object can then be used as input to the NoiseModelFromNoiseProperties
+    class (cirq.devices.noise_properties) to create a NoiseModel that can be used with a simulator.
```
Backticks for docs linking .

95-martin-orion replied:
Done.
